include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -513,6 +515,27 @@ __radix_sort_reorder_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments,
                     }
                     __output_rng[__new_offset_idx] = __in_val;
                 }
+                if (__residual > 0)
+                {
+                    _InputT __in_val{};
+                    ::std::uint32_t __bucket = __radix_states; // greater than any actual radix state
+                    if (__self_lidx < __residual)
+                    {
+                        __in_val = __input_rng[__seg_end + __self_lidx];
+                        __bucket = __get_bucket<(1 << __radix_bits) - 1>(
+                            __order_preserving_cast<__is_ascending>(__in_val), __radix_offset);
+                    }
+                    _OffsetT __new_offset_idx = 0;
+                    for (::std::uint32_t __radix_state_idx = 0; __radix_state_idx < __radix_states; ++__radix_state_idx)
```

Comment:
I'm scratching my head about this for a while, and do not know how to do it best. A function would need to take several parameters, including `peer_contribution`, and I cannot clearly formulate its semantics. Perhaps making it a method of peer helpers would make sense, but then we would duplicate the loop three times, not two :) I thought of creating a local lambda function, but that would change the code sequence (by moving the inner loop before the outer loop), and performance impact would need to be evaluated. What else, making a base class for all peer helpers perhaps? That could   result in additional code changes, and I am not sure if it is worth it.

How, you think, this change could/should be handled when the inner loop is changed to iterate over radix bits?

Reply:
In #731, I moved the loop to the helper classes, but you are correct that such a change duplicates the loop three times and not just two. Perhaps a function in a base class would be the cleanest among the solutions, but that would require relatively large changes to the structure and the benefit is not that strong. Ultimately, it might be better to just leave the loop duplicated as is.

Reply:
The base class is the only one that is tempting to me, but with all the parameters I agree that it is best to leave as is.

Reply:
After some thinking, I concluded the best (from the design standpoint) would be to move this loop to peer helpers similarly to #731, perhaps with some refactoring along the way. Added a TODO for this.